feat(metrics): Add rest of master chartdata metrics

uristdwarf opened the pull request:
This commit finishes adding metric data from master chartsdata. Since all the chartsdata was counters in Prometheus terminology, there was no need to add other types, however a a small refactor was needed to accomodate master metrics and in the future other service metrics. Notably removed was CPU usage and memory, since there are better ways to gather this data, including with a prometheus server itself (which will scrape this data).

This commit deprecates master chartsdata, and is the first step in deprecating the CGI monitoring

uristdwarf committed:
feat(metrics): Add rest of master chartdata metrics

This commit finishes adding metric data from master chartsdata. Since
all the chartsdata was counters in Prometheus terminology, there was no
need to add other types, however a a small refactor was needed to
accomodate master metrics and in the future other service metrics.
Notably removed was CPU usage and memory, since there are better ways to
gather this data, including with a prometheus server itself (which will
scrape this data).

This commit deprecates master chartsdata, and is the first step in
deprecating the CGI monitoring

Signed-off-by: Urmas Rist <urmas@leil.io>

lgsilva3087 reviewed (dismissed):
LGTM.
I see you are almost not using anymore #include "common/platform.h". I don't remember the last decision about it, can you please remaind it.

rolysr commented on src/metrics/utils.h:
```diff
@@ -0,0 +1,11 @@
+#pragma once
```
Consider adding corresponding copyright license to new created files.

ralcolea commented on src/master/chartsdata.cc:
```diff
@@ -18,6 +18,9 @@
  */
 
+
+/* DEPRECATED: Use src/metrics instead */
```
I think this comment is somehow ambiguous. 
If the comment is for file `master/charsdata.h`, the comment should be placed beside the file which is being deprecated. Otherwise, it should be explicitely mentioned the name of file(s) that are being deprecated.

ralcolea commented on src/metrics/master.h:
```diff
@@ -0,0 +1,25 @@
+
+namespace metrics::master {
+
+struct Master {
```
Why use `struct` instead of `class`?

uristdwarf replied:
struct is public by default compared to class, which I generally prefer if I have few/no members that are private and no inheritance is intended for it (i.e it's a simple data structure).

ralcolea commented on src/metrics/master.h:
```diff
@@ -0,0 +1,25 @@
+	Master(std::shared_ptr<prometheus::Registry>& registry);
+
+	// Metric(s)
+	prometheus::Family<prometheus::Counter> *packet_client_counter{nullptr};
```
Maybe `prometheus::Family<prometheus::Counter>` can be defined as a custom data type with a shorter name by using C++ `using ` instruction and then use this data type in all functions.

```cpp
using MetricsCounter = prometheus::Family<prometheus::Counter>;

// Metric(s)
MetricsCounter *packet_client_counter{nullptr};
MetricsCounter *byte_client_counter{nullptr};
MetricsCounter *filesystem_counter{nullptr};
MetricsCounter *chunk_counter{nullptr};
```

ralcolea commented on src/metrics/master.cc:
```diff
@@ -0,0 +1,166 @@
+		// set. Any enum value not used will throw a compile time error.
+		master::Counters start = KEY_START;
+		switch (start) {
+        case KEY_START:
```
Why the alignment is different for `case KEY_START:` ?

ralcolea commented on src/metrics/master.cc:
```diff
@@ -0,0 +1,166 @@
+				{{"filesystem", "operations"}, {"operation", "READ"}},
+			    filesystem_counter);
+			[[fallthrough]];
+        case FS_WRITE:
```
Same here about different alignment. Please check the rest of the function.

ralcolea commented on src/metrics/metrics.cc:
```diff
@@ -64,56 +81,32 @@ prometheus::Family<prometheus::Counter> &setup_family(
-	std::array<Counter, Counter::Key::KEY_END + 1> counters;
 };
-Counters counters;  // NOLINT(cppcoreguidelines-avoid-non-const-global-variables)
+PrometheusMetrics prom_metrics;  // NOLINT(cppcoreguidelines-avoid-non-const-global-variables)
```
Why not `prometheus_metrics` instead of `prom_metrics`? 
I think is more clear.

ralcolea commented on src/metrics/metrics.cc:
```diff
@@ -34,8 +34,25 @@
+#include "metrics/master.h"
 #include "slogger/slogger.h"
 
+// Quick summary on the current implementation from 17/07/24
```
I think in the final version, there should not be implementation details referencing specific dates.

uristdwarf replied:
My worry is that comments go out of date fast, but I guess that one can look at git blame instead.

ralcolea reviewed (dismissed):
Very good job @uristdwarf :+1: :fire: :rocket: 

I added some minor suggestions to improve readability and other aspects, but they are not stoppers for merging.

uristdwarf commented:
> LGTM. I see you are almost not using anymore #include "common/platform.h". I don't remember the last decision about it, can you please remaind it.

For new libraries, I think we are not adding it but for existing/refactored ones we are

ralcolea reviewed (dismissed):
LGTM :+1: :fire: :rocket: 
The only suggestion I have before merging (if not done) is to run `test_prometheus.sh` with `ENABLE_PROMETHEUS = 0` to confirm everything works as expected.

lgsilva3087 reviewed (changes requested):
Requesting changes about bytes sent and received. The call to increment should use the second argument (by default it is 1).

lgsilva3087 commented:
> > LGTM. I see you are almost not using anymore #include "common/platform.h". I don't remember the last decision about it, can you please remaind it.
> 
> For new libraries, I think we are not adding it but for existing/refactored ones we are

Maybe this kind of decisions should be communicated to all devs.

uristdwarf committed:
build: Get Prometheus from vcpkg

Previously it required you to have the library installed on the system.
Using vcpkg, it is now acquired externally if using vcpkg. It will still
try to use system libraries if vcpkg is not being used. This will
statically compile the library into the binary.

Signed-off-by: Urmas Rist <urmas@leil.io>

uristdwarf commented on src/master/matoclserv.cc:
```diff
@@ -5233,6 +5233,7 @@ void matoclserv_read(matoclserventry *eptr) {
 		}
 		eptr->inputpacket.startptr+=i;
 		eptr->inputpacket.bytesleft-=i;
+		metrics::Counter::increment(metrics::master::CLIENT_RX_BYTES);
```
Nice catch, fixed

uristdwarf commented on src/master/matoclserv.cc:
```diff
@@ -5304,13 +5305,14 @@ void matoclserv_write(matoclserventry *eptr) {
 		}
 		pack->startptr+=i;
 		pack->bytesleft-=i;
+		metrics::Counter::increment(metrics::master::CLIENT_TX_BYTES);
```
Also true

uristdwarf commented on src/metrics/master.h:
```diff
@@ -0,0 +1,43 @@
+	CounterFamily *packet_client_counter{nullptr};
+	CounterFamily *byte_client_counter{nullptr};
+	CounterFamily *filesystem_counter{nullptr};
+	CounterFamily *chunk_counter{nullptr};
```
Done. I think it's mostly because C++ library functions are snake_case, and I get confused sometimes

uristdwarf commented on src/metrics/metrics.h:
```diff
@@ -27,35 +27,62 @@
 
+namespace master {
+
+enum Counters {
```
This was more of a pain than I realized, as I used implicit conversion a lot more than expected. Still, I think it's better now because it prevents programmer error from implicitly passing a non-enum to the interface functions.

uristdwarf commented on src/metrics/master.cc:
```diff
@@ -0,0 +1,185 @@
+	// A very hacky way to allow compile time checking if metrics have been
+	// set. Any enum value not used will throw a compile time error.
+	master::Counters start = KEY_START;
+	switch (start) {
```
Not sure how. My idea is that adding enum members will raise compile warnings here, so you don't forget about this place (and try to figure why your metric isn't working)

uristdwarf committed:
test: Add tests for checking master prometheus

Specifically, it checks that the master operations are working

Signed-off-by: Urmas Rist <urmas@leil.io>

uristdwarf commented on tests/test_suites/SanityChecks/test_prometheus.sh:
I've for now only extended them to the operations in metadata, because I'm unsure about the other operations. I've checked the CGI test and all it does is check if the data is populated.

lgsilva3087 reviewed (dismissed):
LGTM

ralcolea commented on tests/test_suites/ShortSystemTests/test_prometheus_metadata.sh:
```diff
@@ -0,0 +1,37 @@
+	if grep -q "READLINK" <<< "$line"; then
+		continue
+	fi
+    # Extract the numeric value from each line
```
Please, check the indentation of lines 32-35. These lines uses spaces instead of tabs.

ralcolea reviewed (dismissed):
Great job @uristdwarf :+1: :fire: :rocket:
I shared some minor suggestions for your consideration, but none of them are stoppers to merge.

However, I had compilation errors when building the project without installing `prometheus-cpp-dev` package. Can you confirm whether this scenario works as expected?

ralcolea reviewed (dismissed):
Great job @uristdwarf :+1: :fire: :rocket:

lgsilva3087 reviewed (approved):
LGTM

ralcolea reviewed (approved):
LGTM :clap:
